fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/domain/SearchQueryRenderer.java

```diff
@@ -0,0 +1,2026 @@
+                 AND EXISTS (
+              SELECT 1
+                FROM Patient_LOGICAL_RESOURCES AS LR1
+          INNER JOIN Patient_STR_VALUES AS P2 ON P2.LOGICAL_RESOURCE_ID = LR1.LOGICAL_RESOURCE_ID
+                 AND P2.PARAMETER_NAME_ID = 1246
+                 AND (P2.STR_VALUE = ?)
+               WHERE LR1.IS_DELETED = 'N'
+                 AND LR1.LOGICAL_RESOURCE_ID = LR0.LOGICAL_RESOURCE_ID)) AS LR
+          INNER JOIN Patient_RESOURCES AS R ON LR.CURRENT_RESOURCE_ID = R.RESOURCE_ID
+            ORDER BY LR.LOGICAL_RESOURCE_ID
+         FETCH FIRST 10 ROWS ONLY
+         */
+
+        // The root query is just the inner distinct piece. The overall query is built by wrapInclude
+        final boolean distinct = true;
+        SelectAdapter select = Select.select(distinct, "R0.RESOURCE_ID", "R0.LOGICAL_RESOURCE_ID", "R0.VERSION_ID", "R0.LAST_UPDATED", "R0.IS_DELETED", "LR0.LOGICAL_ID");
+        return new QueryData(select, null, null, rootResourceType, 0);
+    }
+
+    @Override
+    public QueryData wrapInclude(QueryData query) {
+        // Need to join the RESOURCES table again to get the DATA column after the DISTINCT.
+        final String lrAlias = "LR";
+        final String rAlias = "R";
+        final String rTable = query.getResourceType() + "_RESOURCES";
+        SelectAdapter select = Select.select("LR.RESOURCE_ID", "LR.LOGICAL_RESOURCE_ID", "LR.VERSION_ID", "LR.LAST_UPDATED", "LR.IS_DELETED", "R.DATA", "LR.LOGICAL_ID");
+        select.from(query.getQuery().build(), alias(lrAlias))
+            .innerJoin(rTable, alias(rAlias), on(lrAlias, "RESOURCE_ID").eq(rAlias, "RESOURCE_ID"));
+        return new QueryData(select, lrAlias, null, query.getResourceType(), 0);
+    }
+
+    @Override
+    public QueryData sortRoot(String rootResourceType) {
+        final String xxLogicalResources = resourceLogicalResources(rootResourceType);
+        final String lrAliasName = "LR0";
+
+        // The core data query joining together the logical resources table. Query
+        // parameters are bolted on as exists statements in the WHERE clause. The final
+        // query is constructed when joinResources is called.
+        SelectAdapter select = Select.select("LR0.CURRENT_RESOURCE_ID");
+        select.from(xxLogicalResources, alias(lrAliasName))
+            .where(lrAliasName, IS_DELETED).eq().literal("N");
+
+        // We need to group the sort parameters to address any duplicates
+        select.from().groupBy("LR0.CURRENT_RESOURCE_ID");
+        return new QueryData(select, lrAliasName, null, rootResourceType, 0);
+    }
+
+    /**
+     * Get the filter predicate for the given token query parameter
+     * @param queryParm the token query parameter
+     * @param paramAlias the alias used for the token values table
+     * @throws FHIRPersistenceException
+     */
+    protected WhereFragment getTokenFilter(QueryParameter queryParm, String paramAlias) throws FHIRPersistenceException {
```

Comment:
Do you need both getTokenFilter() and getComplexTokenFilter()? Down on line 1353 you only call getTokenFilter() if the operator is EQ, but it looks like the same code that's executed in this method would get executed if you called getComplexTokenFilter() instead with an operator of EQ. Just wondering if you could always just call getComplexTokenFilter() (and maybe rename it to getTokenFilter()).

Reply:
Thanks Mike, good observation. Lots of refactoring in this error as I was getting the new query builder to work. Have cleaned up as you suggested and it still passes all the tests.